spec/lang/step.md

```diff
@@ -230,36 +268,48 @@ impl<M: Memory> Machine<M> {
 
 ```rust
 impl<M: Memory> Machine<M> {
-    fn eval_place(&mut self, PlaceExpr::Field { root, field }: PlaceExpr) -> NdResult<Place<M>> {
-        let ty = root.check_wf::<M>(self.cur_frame().func.locals, self.prog).unwrap().ty; // FIXME avoid a second traversal of `root`
-        let root = self.eval_place(root)?;
-        let offset = match ty {
-            Type::Tuple { fields, .. } => fields[field].0,
-            Type::Union { fields, .. } => fields[field].0,
+    fn eval_place(&mut self, PlaceExpr::Field { root, field }: PlaceExpr) -> NdResult<(Place<M>, PlaceType)> {
+        let (root, ptype) = self.eval_place(root)?;
+        let (offset, field_ty) = match ptype.ty {
+            Type::Tuple { fields, .. } => fields[field],
+            Type::Union { fields, .. } => fields[field],
             _ => panic!("field projection on non-projectable type"),
         };
-        assert!(offset <= ty.size::<M>());
-        ret(self.ptr_offset_inbounds(root, offset.bytes())?)
+        assert!(offset <= ptype.ty.size::<M>());
+
+        let place = self.ptr_offset_inbounds(root, offset.bytes())?;
+        let ptype = PlaceType {
+            align: ptype.align.restrict_for_offset(offset),
+            ty: field_ty,
+        };
+        
+        ret((place, ptype))
     }
 
-    fn eval_place(&mut self, PlaceExpr::Index { root, index }: PlaceExpr) -> NdResult<Place<M>> {
-        let ty = root.check_wf::<M>(self.cur_frame().func.locals, self.prog).unwrap().ty; // FIXME avoid a second traversal of `root`
-        let root = self.eval_place(root)?;
-        let Value::Int(index) = self.eval_value(index)? else {
+    fn eval_place(&mut self, PlaceExpr::Index { root, index }: PlaceExpr) -> NdResult<(Place<M>, PlaceType)> {
+        let (root, ptype) = self.eval_place(root)?;
+        let (Value::Int(index), _) = self.eval_value(index)? else {
             panic!("non-integer operand for array index")
         };
-        let offset = match ty {
+        let (offset, field_ty) = match ptype.ty {
             Type::Array { elem, count } => {
                 if index >= 0 && index < count {
-                    index * elem.size::<M>()
+                    (index * elem.size::<M>(), elem)
                 } else {
                     throw_ub!("out-of-bounds array access");
                 }
             }
             _ => panic!("index projection on non-indexable type"),
         };
-        assert!(offset <= ty.size::<M>());
-        ret(self.ptr_offset_inbounds(root, offset.bytes())?)
+        assert!(offset <= ptype.ty.size::<M>());
+
+        let place = self.ptr_offset_inbounds(root, offset.bytes())?;
+        let ptype = PlaceType {
+            align: ptype.align.restrict_for_offset(field_ty.size::<M>()),
```

Comment:
```suggestion
            // We do *not* use `offset` here since that is only dynamically known.
            align: ptype.align.restrict_for_offset(field_ty.size::<M>()),
```